platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/cli/BuildEnvironmentConfigurationConverter.java

```diff
@@ -31,61 +32,78 @@
+                                           LayoutToPropertiesConverter layoutToPropertiesConverter,
+                                           StartParameterConverter startParameterConverter,
+                                           BuildOptionBackedConverter<DaemonParameters> daemonParametersConverter,
+                                           BuildOptionBackedConverter<DaemonJvmToolchainSpec> daemonJvmToolchainCriteriaConverter,
+                                           FileCollectionFactory fileCollectionFactory,
+                                           PropertyFactory propertyFactory) {
         this.initialPropertiesConverter = initialPropertiesConverter;
         this.buildLayoutConverter = buildLayoutConverter;
         this.layoutToPropertiesConverter = layoutToPropertiesConverter;
         this.startParameterConverter = startParameterConverter;
         this.daemonParametersConverter = daemonParametersConverter;
+        this.daemonJvmToolchainCriteriaConverter = daemonJvmToolchainCriteriaConverter;
         this.fileCollectionFactory = fileCollectionFactory;
+        this.propertyFactory = propertyFactory;
     }
 
-    public ParametersConverter(BuildLayoutFactory buildLayoutFactory, FileCollectionFactory fileCollectionFactory) {
+    public BuildEnvironmentConfigurationConverter(BuildLayoutFactory buildLayoutFactory, FileCollectionFactory fileCollectionFactory, PropertyFactory propertyFactory) {
         this(new InitialPropertiesConverter(),
             new BuildLayoutConverter(),
             new LayoutToPropertiesConverter(buildLayoutFactory),
             new StartParameterConverter(),
             new BuildOptionBackedConverter<>(new DaemonBuildOptions()),
-            fileCollectionFactory);
+            new BuildOptionBackedConverter<>(new DaemonJvmToolchainCriteriaOptions()),
+            fileCollectionFactory,
+            propertyFactory);
     }
 
-    public Parameters convert(ParsedCommandLine args, @Nullable File currentDir) throws CommandLineArgumentException {
+    public Parameters convertParameters(ParsedCommandLine args, @Nullable File currentDir) throws CommandLineArgumentException {
         InitialProperties initialProperties = initialPropertiesConverter.convert(args);
         BuildLayoutResult buildLayout = buildLayoutConverter.convert(initialProperties, args, currentDir);
         AllProperties properties = layoutToPropertiesConverter.convert(initialProperties, buildLayout);
-
         StartParameterInternal startParameter = new StartParameterInternal();
         startParameterConverter.convert(args, buildLayout, properties, startParameter);
 
         DaemonParameters daemonParameters = new DaemonParameters(buildLayout, fileCollectionFactory, properties.getRequestedSystemProperties());
-        daemonParametersConverter.convert(args, properties, daemonParameters);
+        daemonParametersConverter.convert(args, properties.getProperties(), daemonParameters);
```

Comment:
I am confused why we are instantiating another `DaemonJvmToolchainCriteriaOptions` above and converting here. 
Shouldn't our build properties be present in `AllProperties#getBuildProperties` already?

I might be missing something. But I think we want to try to keep these sets of properties separate

Reply:
This is exactly how the existing `DaemonBuildOptions` works you have the `BuildOptionBackedConverter<T>` that requires a `BuildOptionSet` instance and then  also on `LayoutToPropertiesConverter` is using the instance instead of accessing the static `options` field. I do agree that for the last one we can use directly the static method instead created a new instance. 

 What is important here and the reason why this class was renaming to `BuildEnvironmentConfigurationConverter` is because we want to have two converters one for `convertParameters` (existing one) and `convertJvmToolchainCriteria` (new one). Its important not mixing them together because the second one can potentially fail to convert but this should be ignored in case the `updateDaemonJvm` task was executed since instead the JAVA_HOME environment variable will be used to update the build properties with specified criteria. This is important since we want to be sure given an invalid criteria or unable to find any local toolchain matching the criteria, the build properties can be updated properly for next build to be taken into consideration.
 
 More context about this can be found also [here](https://github.com/gradle/gradle/pull/27898#discussion_r1481778145)